File: contact_list.rb
class Person < Struct.new(:name)
  # Say hi to person
  def say_hi_to(person)
    "Hi #{person.name}, my name is #{name}."
  end
end

class ContactList
  attr_reader :contacts, :owner

  def initialize(person)
    @owner    = person
    @contacts = []
  end
  # Returns array with contacts who's name start with name
  def contacts_by_name(name)
    contacts.select { |p| p.name.start_with?(name) }
  end
  # Returns first match for name
  def contact_by_name(name)
    contacts_by_name(name).first
  end
  # Returns true if person is in contact list
  def in_contact_list?(person)
    contact_by_name(person.name) != nil
  end
  # Add person to contact list, returns this
  def add_contact(person)
    @contacts << person
    self
  end
end
# Data seeds
wille = Person.new('Wille')
becka = Person.new('Becka')
jacob = Person.new('Jacob')
vigge = Person.new('Vigge')
sigge = Person.new('Sigge')
# Add contacts
list = ContactList.new(wille)
list.add_contact(becka)
list.add_contact(jacob)
list.add_contact(vigge)
list.add_contact(sigge)

puts list.contacts_by_name('Ja')[0].name # Find contacts who's name start with 'Ja'
puts wille.say_hi_to(becka)              # Wille say hi to Becka

File: contact-list.js
var Person = function(name) {
  this.self  = this;
  self.name  = name;
  // Say hi to person
  self.sayHiTo = function(person) {
    return 'Hi ' + person.name + ', my name is ' + self.name + '.';
  };
};

var ContactList = function(person) {
  this.self     = this;
  self.owner    = person;
  self.contacts = [];
  // Returns array with contacts who's name start with name
  self.getContactsByName = function(name) {
    return self.contacts.filter(function(person) {
      return person.name.startsWith(name);
    });
  };
  // Returns first match for name
  self.getContactByName = function(name) {
    return self.getContactsByName(name)[0];
  };
  // Returns true if person is in contact list
  self.inContactList = function(person) {
    return self.getContactByName(person.name) !== undefined;
  };
  // Add person to contact list, returns this
  self.addContact = function(person) {
    self.contacts.push(person);
    return self;
  };
};

// Data seeds
var wille  = new Person('Wille');
var becka  = new Person('Becka');
var philip = new Person('Philip');
var jacob  = new Person('Jacob');
var janush = new Person('Janush');
// Add contacts
contacts = new ContactList(wille);
contacts.addContact(becka);
contacts.addContact(philip);
contacts.addContact(jacob);
contacts.addContact(janush);

console.log(contacts.getContactsByName('Ja')[0].name); // Find contacts who's name start with 'Ja'
console.log(wille.sayHiTo(becka));                     // Wille say hi to Becka
